flocker/common/test/test_algebraic.py

```diff
@@ -0,0 +1,117 @@
+# -*- test-case-name: flocker.common.test.test_algebraic -*-
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Tests for ``flocker.common.algebraic``
+"""
+
+from pyrsistent import (
+    PClass, field,
+    InvariantException,
+)
+from hypothesis import given, strategies as st, assume
+
+from twisted.trial.unittest import SynchronousTestCase
+
+from twisted.python.constants import Names, NamedConstant
+
+
+from ..algebraic import StateInvariant
+
+
+class States(Names):
+    ALLOWED = NamedConstant()
+    WITH_ATTRIBUTE = NamedConstant()
+    WITH_TWO_ATTRIBUTES = NamedConstant()
+    DISALLOWED = NamedConstant()
+
+
+class AlgebraicType(PClass):
+    state = field(mandatory=True)
+    one = field(bool)
+    two = field(bool)
+
+    __invariant__ = StateInvariant(
+        allowed_states={
+            States.ALLOWED,
+            States.WITH_ATTRIBUTE,
+            States.WITH_TWO_ATTRIBUTES,
+        },
+        expected_attributes={
+            'one': {States.WITH_ATTRIBUTE, States.WITH_TWO_ATTRIBUTES},
+            'two': {States.WITH_TWO_ATTRIBUTES},
+        },
+    )
+
+ARGS_STRATEGEY = st.fixed_dictionaries({
+    'state': st.sampled_from(AlgebraicType.__invariant__.allowed_states),
+    'one': st.booleans(),
+    'two': st.booleans(),
+})
+
+
+class MakeStateInvariantTests(SynchronousTestCase):
+
+    @given(
+        args=ARGS_STRATEGEY,
+        extra_attribute=st.sampled_from(
+            AlgebraicType.__invariant__.expected_attributes
+        ),
+    )
+    def test_extra_attributes(self, args, extra_attribute):
+        for attribute, states in (
+                AlgebraicType.__invariant__.expected_attributes.items()
+        ):
+            if attribute == extra_attribute:
+                assume(args['state'] not in states)
+                continue
+            if args['state'] not in states:
+                del args[attribute]
```

Comment:
It took me a while to figure out what this `for` loop was doing. I think it's trying to generate a state with an extra argument that isn't valid in that state.

Rather than doing post-processing in the tests, it's better to have some helpers to generate the right thing. Sketch follows:

```python
states = st.sampled_from(AlgebraicType.__invariant__.allowed_states)
valid_attributes = st.sampled_from(AlgebraicType.__invariant__.expected_attributes)

def args_for_state(state):
    # XXX: This would be simpler if `expected_attributes` were inverted.
    args = {'state': st.just(state)}
    for arg, attrs in AlgebraicType.__invariant__.expected_attributes.items():
        if state in attrs:
            args[arg] = st.booleans()
    return st.fixed_dictionaries(args)

valid_arguments = states.flatmap(args_for_state)

@given(
  args=valid_arguments,
  extra_attribute=valid_attributes,
  extra_arg=st.booleans(),
)
def test_extra_attributes(self, args, extra_attribute, extra_arg):
    assume(extra_attribute not in args)
    args[extra_attribute] = extra_arg
    exc = ...
```

another approach might be to use the new, semi-documented `choices` strategy (http://www.drmaciver.com/2015/11/let-hypothesis-making-your-choices-for-you/)

```python
@given(
  args=valid_arguments,
  extra_arg=st.booleans(),
  choice=st.choices(),
)
def test_extra_attributes(self, args, extra_arg, choice):
    extra_attribute = choice(set(AlgebraicType.__invariant__.allowed_states.keys()) - set(args.keys()))
    args[extra_attribute] = extra_arg
    ...
```